Fix Android Studio support

jkmassel opened the pull request:
Works around https://github.com/gradle/gradle/issues/10483 by resolving the compiler paths at the start of the script, then passing them where they’re needed.

This is only coming up now because we’ve adopted JDK21, which seems to be related to the issue at hand (at least on a Mac).

### To Test:
Before running any tests, run `git clean -fdX` from the root of the repo, and ensure you've run `make setup-rust-android-targets` successfully.

**Android Studio**
1. Open the project in Android Studio
2. Ensure the initial Gradle Sync works
3. Run the `ApiUrlDiscoveryTest` suite from within Android Studio (in `api/kotlin/integrationTest/tests/ApiUrlDiscoveryTest.kt`) to ensure that all the JNI stuff is working.
4. Try building the `example.composeApp` configuration – it should work.

**Gradle**
1. Run `./native/kotlin/gradlew -p native/kotlin compileKotlin`
2. Note that the output is printed as the command runs
3. Run the `ApiUrlDiscoveryTest` suite using `./native/kotlin/gradlew -p native/kotlin :api:kotlin:integrationTest --tests rs.wordpress.api.kotlin.ApiUrlDiscoveryTest` to ensure that all the JNI stuff is working.
4. Build the `example.composeApp` configuration using `./native/kotlin/gradlew -p native/kotlin :example:composeApp:assembleDebug` to ensure that the Android bits are working.

jkmassel commented on native/kotlin/api/kotlin/build.gradle.kts:
```diff
@@ -76,50 +76,24 @@ dependencies {
     implementation(libs.kotlinx.coroutines.core)
 }
 
+val generatedSourceDirectory = rootProject.layout.buildDirectory.get()
```
Building up paths this way is better for cross-platform support (I can't see any reason this _shouldn't_ work on Windows, but at the moment I'm pretty sure it doesn't)

adalpari commented:
After running

```
adalpari@Adalbertos-MacBook-Pro-2 wordpress-rs % git clean -fdX
Removing .DS_Store
Removing native/.DS_Store
Removing native/kotlin/.claude/
Removing native/kotlin/.gradle/
Removing native/kotlin/.idea/
Removing native/kotlin/.kotlin/
Removing native/kotlin/api/android/build/
Removing native/kotlin/api/kotlin/build/
Removing native/kotlin/build/
Removing native/kotlin/example/composeApp/build/
Removing native/kotlin/local.properties
Removing target/
Removing test_credentials.json
adalpari@Adalbertos-MacBook-Pro-2 wordpress-rs % make setup-rust-android-targets
rustup target add \
		x86_64-linux-android \
		i686-linux-android \
		armv7-linux-androideabi \
		aarch64-linux-android
info: component 'rust-std' for target 'x86_64-linux-android' is up to date
info: component 'rust-std' for target 'i686-linux-android' is up to date
info: component 'rust-std' for target 'armv7-linux-androideabi' is up to date
info: component 'rust-std' for target 'aarch64-linux-android' is up to date
```


Now, when I open AS, the initial sync fails, but it did work before...

> Execution failed for task ':cargoBuildLibraryRelease'

```
Type-safe project accessors is an incubating feature.
> Task :prepareKotlinBuildScriptModel UP-TO-DATE
> Task :api:kotlin:checkKotlinGradlePluginConfigurationErrors SKIPPED
> Task :example:composeApp:convertXmlValueResourcesForAndroidDebug NO-SOURCE
> Task :api:kotlin:processResources NO-SOURCE
> Task :example:composeApp:convertXmlValueResourcesForAndroidInstrumentedTest NO-SOURCE
> Task :example:composeApp:convertXmlValueResourcesForAndroidInstrumentedTestDebug NO-SOURCE
> Task :example:composeApp:convertXmlValueResourcesForAndroidMain NO-SOURCE
> Task :example:composeApp:convertXmlValueResourcesForAndroidRelease NO-SOURCE
> Task :example:composeApp:convertXmlValueResourcesForAndroidUnitTest NO-SOURCE
> Task :cargoBuildLibraryRelease FAILED
> Task :example:composeApp:convertXmlValueResourcesForAndroidUnitTestDebug NO-SOURCE
> Task :example:composeApp:convertXmlValueResourcesForAndroidUnitTestRelease NO-SOURCE
> Task :example:composeApp:convertXmlValueResourcesForCommonMain NO-SOURCE
> Task :example:composeApp:convertXmlValueResourcesForCommonTest NO-SOURCE
> Task :example:composeApp:convertXmlValueResourcesForDesktopMain NO-SOURCE
> Task :example:composeApp:convertXmlValueResourcesForDesktopTest NO-SOURCE
> Task :example:composeApp:copyNonXmlValueResourcesForAndroidDebug NO-SOURCE
> Task :example:composeApp:copyNonXmlValueResourcesForAndroidInstrumentedTest NO-SOURCE
> Task :example:composeApp:copyNonXmlValueResourcesForAndroidInstrumentedTestDebug NO-SOURCE
> Task :example:composeApp:copyNonXmlValueResourcesForAndroidMain NO-SOURCE
> Task :example:composeApp:copyNonXmlValueResourcesForAndroidRelease NO-SOURCE
> Task :example:composeApp:copyNonXmlValueResourcesForAndroidUnitTest NO-SOURCE
> Task :example:composeApp:copyNonXmlValueResourcesForAndroidUnitTestDebug NO-SOURCE
> Task :example:composeApp:copyNonXmlValueResourcesForAndroidUnitTestRelease NO-SOURCE
> Task :example:composeApp:copyNonXmlValueResourcesForCommonMain
> Task :example:composeApp:copyNonXmlValueResourcesForCommonTest NO-SOURCE
> Task :example:composeApp:copyNonXmlValueResourcesForDesktopMain NO-SOURCE
> Task :example:composeApp:copyNonXmlValueResourcesForDesktopTest NO-SOURCE
> Task :example:composeApp:generateComposeResClass
> Task :example:composeApp:prepareComposeResourcesTaskForAndroidDebug NO-SOURCE
> Task :example:composeApp:generateResourceAccessorsForAndroidDebug NO-SOURCE
> Task :example:composeApp:prepareComposeResourcesTaskForAndroidInstrumentedTest NO-SOURCE
> Task :example:composeApp:generateResourceAccessorsForAndroidInstrumentedTest NO-SOURCE
> Task :example:composeApp:prepareComposeResourcesTaskForAndroidInstrumentedTestDebug NO-SOURCE
> Task :example:composeApp:generateResourceAccessorsForAndroidInstrumentedTestDebug NO-SOURCE
> Task :example:composeApp:prepareComposeResourcesTaskForAndroidMain NO-SOURCE
> Task :example:composeApp:generateResourceAccessorsForAndroidMain NO-SOURCE
> Task :example:composeApp:prepareComposeResourcesTaskForAndroidRelease NO-SOURCE
> Task :example:composeApp:generateResourceAccessorsForAndroidRelease NO-SOURCE
> Task :example:composeApp:prepareComposeResourcesTaskForAndroidUnitTest NO-SOURCE
> Task :example:composeApp:generateResourceAccessorsForAndroidUnitTest NO-SOURCE
> Task :example:composeApp:prepareComposeResourcesTaskForAndroidUnitTestDebug NO-SOURCE
> Task :example:composeApp:generateResourceAccessorsForAndroidUnitTestDebug NO-SOURCE
> Task :example:composeApp:prepareComposeResourcesTaskForAndroidUnitTestRelease NO-SOURCE
> Task :example:composeApp:generateResourceAccessorsForAndroidUnitTestRelease NO-SOURCE
> Task :example:composeApp:prepareComposeResourcesTaskForCommonMain UP-TO-DATE
> Task :example:composeApp:generateResourceAccessorsForCommonMain
> Task :example:composeApp:prepareComposeResourcesTaskForCommonTest NO-SOURCE
> Task :example:composeApp:generateResourceAccessorsForCommonTest NO-SOURCE
> Task :example:composeApp:prepareComposeResourcesTaskForDesktopMain NO-SOURCE
> Task :example:composeApp:generateResourceAccessorsForDesktopMain NO-SOURCE
> Task :example:composeApp:prepareComposeResourcesTaskForDesktopTest NO-SOURCE
> Task :example:composeApp:generateResourceAccessorsForDesktopTest NO-SOURCE

[Incubating] Problems report is available at: file:///Users/adalpari/Workspace/wordpress-rs/native/kotlin/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.14.3/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
5 actionable tasks: 4 executed, 1 up-to-date

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':cargoBuildLibraryRelease'.
> A problem occurred starting process 'command 'cargo''

* Try:
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':cargoBuildLibraryRelease'.
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:130)
	at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:293)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:128)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:116)
	at org.gradle.api.internal.tasks.execution.ProblemsTaskPathTrackingTaskExecuter.execute(ProblemsTaskPathTrackingTaskExecuter.java:41)
	at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
	at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:74)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:210)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:205)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:67)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:60)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:167)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:60)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:54)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
	at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:42)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:331)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:318)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.lambda$execute$0(DefaultTaskExecutionGraph.java:314)
	at org.gradle.internal.operations.CurrentBuildOperationRef.with(CurrentBuildOperationRef.java:85)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:314)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:303)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:459)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:376)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:48)
Caused by: org.gradle.process.internal.ExecException: A problem occurred starting process 'command 'cargo''
	at org.gradle.process.internal.DefaultExecHandle.execExceptionFor(DefaultExecHandle.java:241)
	at org.gradle.process.internal.DefaultExecHandle.setEndStateInfo(DefaultExecHandle.java:218)
	at org.gradle.process.internal.DefaultExecHandle.failed(DefaultExecHandle.java:396)
	at org.gradle.process.internal.ExecHandleRunner.lambda$run$3(ExecHandleRunner.java:102)
	at org.gradle.internal.operations.CurrentBuildOperationRef.with(CurrentBuildOperationRef.java:85)
	at org.gradle.process.internal.ExecHandleRunner.run(ExecHandleRunner.java:101)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:48)
Caused by: net.rubygrapefruit.platform.NativeException: Could not start 'cargo'
	at net.rubygrapefruit.platform.internal.DefaultProcessLauncher.start(DefaultProcessLauncher.java:27)
	at net.rubygrapefruit.platform.internal.WrapperProcessLauncher.start(WrapperProcessLauncher.java:36)
	at org.gradle.process.internal.ExecHandleRunner.startProcess(ExecHandleRunner.java:122)
	at org.gradle.process.internal.ExecHandleRunner.lambda$run$0(ExecHandleRunner.java:80)
	at org.gradle.internal.operations.CurrentBuildOperationRef.with(CurrentBuildOperationRef.java:85)
	at org.gradle.process.internal.ExecHandleRunner.run(ExecHandleRunner.java:79)
	... 2 more
Caused by: java.io.IOException: Cannot run program "cargo" (in directory "/Users/adalpari/Workspace/wordpress-rs"): error=2, No such file or directory
	at net.rubygrapefruit.platform.internal.DefaultProcessLauncher.start(DefaultProcessLauncher.java:25)
	... 7 more
Caused by: java.io.IOException: error=2, No such file or directory
	... 8 more


BUILD FAILED in 288ms
```

jkmassel committed:
Fix Android Studio JDK 21 / Rust issues

Works around https://github.com/gradle/gradle/issues/10483 by resolving the compiler paths at the start of the script, then passing them where they’re needed.

adalpari reviewed (approved):
Now it works as a charm!! Thank you for the changes 🙂
